backend/src/s9pk/docker.rs

```diff
@@ -0,0 +1,94 @@
+use std::borrow::Cow;
+use std::collections::BTreeSet;
+use std::io::SeekFrom;
+use std::path::Path;
+
+use color_eyre::eyre::eyre;
+use futures::{FutureExt, TryStreamExt};
+use serde::{Deserialize, Serialize};
+use tokio::io::{AsyncRead, AsyncSeek, AsyncSeekExt};
+use tokio_tar::{Archive, Entry};
+
+use crate::util::io::from_cbor_async_reader;
+use crate::{Error, ErrorKind, ARCH};
+
+#[derive(Default, Deserialize, Serialize)]
+#[serde(rename_all = "kebab-case")]
+pub struct DockerMultiArch {
+    pub default: String,
+    pub available: BTreeSet<String>,
+}
+
+#[pin_project::pin_project(project = DockerReaderProject)]
+pub enum DockerReader<R: AsyncRead + Unpin> {
+    SingleArch(#[pin] R),
+    MultiArch(#[pin] Entry<Archive<R>>),
+}
+impl<R: AsyncRead + AsyncSeek + Unpin + Send + Sync> DockerReader<R> {
+    pub async fn new(mut rdr: R) -> Result<Self, Error> {
+        let arch = if let Some(multiarch) = tokio_tar::Archive::new(&mut rdr)
+            .entries()?
+            .try_filter_map(|e| {
+                async move {
+                    Ok(if &*e.path()? == Path::new("multiarch.cbor") {
+                        Some(e)
+                    } else {
+                        None
+                    })
+                }
+                .boxed()
+            })
+            .try_next()
+            .await?
+        {
+            let multiarch: DockerMultiArch = from_cbor_async_reader(multiarch).await?;
+            Some(if multiarch.available.contains(&**ARCH) {
+                Cow::Borrowed(&**ARCH)
+            } else {
+                Cow::Owned(multiarch.default)
+            })
+        } else {
+            None
+        };
+        rdr.seek(SeekFrom::Start(0)).await?;
+        if let Some(arch) = arch {
+            if let Some(image) = tokio_tar::Archive::new(rdr)
+                .entries()?
+                .try_filter_map(|e| {
+                    let arch = arch.clone();
+                    async move {
+                        Ok(if &*e.path()? == Path::new(&format!("{}.tar", arch)) {
+                            Some(e)
+                        } else {
+                            None
+                        })
+                    }
+                    .boxed()
+                })
+                .try_next()
+                .await?
+            {
+                Ok(Self::MultiArch(image))
+            } else {
+                Err(Error::new(
```

Comment:
:question: Is this running the packing or the running of the system. If it is running, how would it be possible that this error would have been hit, or is this an error that should never have been called.

Reply:
this is the running of the system. and it should never be hit